content/en/blog/2023/2023-03-08-scheduler-plugin-reusing/index.md

```diff
@@ -0,0 +1,124 @@
+
+The dependency to the composite does require some understanding of the composite code, and there may be some introduced fragility with having that dependency. For example, looking into the _NodeResourceFit_ plugin internals reveals that it utilizes the [framework.CycleState](https://github.com/kubernetes/kubernetes/blob/2f977fd8c400971deed365dcc437e519bce475f0/pkg/scheduler/framework/cycle_state.go#L42-L47) during running of other Plugin interfaces. Luckily during running of the `ScorePlugin` interface parts no such state is utilized, so in this case it wasn't necessary to add and run those other plugin interfaces. However should such a thing change in the composite plugin, then running of the composite using plugin might also require changes.
+
+### More efficient and flexible alternative, plugin pipelining
+
+Going through the exercise of building a new plugin via composition and realizing the caveats of such reuse raised some questions related to if and how things could be better in an ideal world. Such as:
+
+* Could a plugin utilize scores from multiple other plugins?
+* Could multiple new plugins utilize scores from the same plugin, without running the same scoring logic N times?
+* Could the dependency to the reused plugin be made less fragile, reducing the coupling?
+* Could reuse be achieved without duplicated args and configuring same algorithms the same way, twice (or more)?
+* Could things be configured more dynamically instead of hard-coding to the sources one plugin uses another plugin?
+
+These questions had the author thinking about finding a better solution than just composition. Also the documentation of the [framework.CycleState](https://github.com/kubernetes/kubernetes/blob/2f977fd8c400971deed365dcc437e519bce475f0/pkg/scheduler/framework/cycle_state.go#L42-L47) was looking rather tempting.
+
+> **Warning**
+> The following section is speculative and based on proof-of-concepting by the author. It doesn't represent what you can do with the Kubernetes scheduler at the time of writing. Some day the concept below might be part of Kubernetes, should the PoC develop into a successful KEP. For now, this is merely a scheduling-sig discussion item which shows the potential of the CycleState.
+
+Ideally, a new plugin could ask for getting the scores from another plugin, or multiple plugins, without rerunning them. The tricky thing is, that plugins run highly parallel in order to execute faster. Thus proper synchronization would be needed. Also it would be ideal, if no changes to existing plugins would be required, and the framework would provide for plugins in need of reusing other plugins' results.
+
+This got the author thinking, that if the `framework.CycleState` object could contain a channel for the scores, and if the framework would create such a channel when it is needed, then the parallel running plugins which are interested in score reuse, could wait for the scores to appear from such "ScoreChannels".
```

Comment:
I'm trying to understanding why we need a channel. If we place the composite plugin in the end (or at least place after their dependent plugins), and then previous plugins write their {pluginName: score} to the CycleState, then I'd assume the composite plugin can reuse those pre-calculated scores already?

Reply:
The channel is only needed for concurrency, and it isn't really necessary at all if one does a composite plugin which reruns the logic of the other plugin. If you rerun the whole other plugin in a composite plugin, you get its score that way, then there's no need to fetch anything from cycle state. The code is reused by running it again, but there is no result reuse.

The point of the channel usage is to not do a composite, and instead run the source plugin only once as it is, and then reuse its result score later. As mentioned in the quoted paragraphs, the score plugins run parallel, and to be able to reuse scores from another plugin, one needs to synchronize. Using a channel is the easy way. The plugin which is in the receiving end waits for the score to appear from the channel. That appearance ensures that the reused plugin has completed processing the score. One could perhaps alternatively poll something in the cycle state object, but that wouldn't be tidy.

You could of course have dedicated processing steps, like "postScore" which runs after "score" and "postPostScore" which runs after "postScore", and then you'd be able to reuse the results from prior steps which could be stored in the cycle state, but why do a bigger framework change like that, if one can do a simple channel solution which allows for a generic solution (unlimited chain).